crates/sui-core/src/authority_active/checkpoint_driver/mod.rs

```diff
@@ -0,0 +1,551 @@
+                    // A byzantine authority!
+                    println!("Sync Error: Incorrect contents returned");
+                    continue;
+                }
+
+                return Ok(contents);
+            }
+            Ok(resp) => {
+                println!("Sync Error: Unexpected answer: {:?}", resp);
+            }
+            Err(err) => {
+                println!("Sync Error: peer error: {:?}", err);
+            }
+        }
+    }
+}
+
+/// Picks other authorities at random and constructs checkpoint fragments
+/// that are submitted to consensus. The process terminates when a future
+/// checkpoint is downloaded
+pub async fn diff_proposals<A>(
+    _active_authority: &ActiveAuthority<A>,
+    checkpoint_db: Arc<Mutex<CheckpointStore>>,
+    _my_proposal: &CheckpointProposal,
+    _proposals: Vec<(AuthorityName, SignedCheckpointProposal)>,
+) where
+    A: AuthorityAPI + Send + Sync + 'static + Clone,
+{
+    // Pick another authority, get their proposal, and submit it to consensus
+    // Exit when we have a checkpoint proposal.
+
+    let mut available_authorities: BTreeSet<_> = _proposals.iter().map(|(auth, _)| *auth).collect();
+    available_authorities.remove(&_active_authority.state.name); // remove ourselves
+    let mut fragments_num = 0;
+
+    loop {
+        let next_checkpoint_sequence_number: u64 = checkpoint_db.lock().next_checkpoint();
+        if next_checkpoint_sequence_number > *_my_proposal.proposal.0.checkpoint.sequence_number() {
+            // Our work here is done, we have progressed past the checkpoint for which we were given a proposal.
+            // Our DB has been updated (presumably by consensus) with the sought information (a checkpoint
+            // for this sequence number)
+            break;
+        }
+
+        // We have ran out of authorities?
+        if available_authorities.is_empty() {
+            // We have created as many fragments as possible, so exit.
+            break;
+        }
+
+        let random_authority = _active_authority.state.committee.sample();
+        if available_authorities.remove(random_authority) {
+            // Get a client
+            let client = _active_authority.net.authority_clients[random_authority].clone();
+
+            if let Ok(response) = client
+                .handle_checkpoint(CheckpointRequest::latest(true))
+                .await
+            {
+                if let AuthorityCheckpointInfo::Proposal { current, previous } = &response.info {
+                    // Check if there is a latest checkpoint
+                    if let AuthenticatedCheckpoint::Certified(prev) = previous {
+                        if prev.checkpoint.sequence_number > next_checkpoint_sequence_number {
+                            // We are now way behind, return
+                            return;
+                        }
+                    }
+
+                    // For some reason the proposal is empty?
```

Comment:
This can happen for epoch change, right?

Reply:
I do not think that should ever happen.